src/shell/signals.c

```diff
@@ -29,6 +29,42 @@
 #include "internal.h"
 #include "builtins.h"
 
+static void init_signal_cb (int signum)
+{
+    /* 128 + signum is standard bash exit code */
+    shell_die (128+signum, "signal %d sent before tasks started", signum);
+}
+
+/* Rare race, signal received before all tasks started.
+ * Hypothetically could be seen with something like:
+ *
+ * jobid=`flux job submit --wait-event=start foo.sh`
+ * flux job raise --type=foo --severity=0 $jobid
+ *
+ * i.e. raise sends SIGTERM sent to job/shell immediately after
+ * starting and before tasks setup.  This should not be seen under
+ * normal operating circumstances.
+ *
+ * This race can lead to unexpected errors due to the signal
+ * interrupting initial setup.  It is hard to predict this situation's
+ * behavior, so we elect to trap these signals and exit, rather than
+ * forward to tasks.
```

Comment:
In the commit message

> Problem: If a trapped signal (SIGTERM, SIGINT, SIGALRM) is sent
> before all shell tasks have been launched, unexpected behavior could
> happen due to signal interruption of task setup (e.g. a write(2)
> call could be interrupted and return EINTR, etc.).

This is a guess at this point so IMO should be left out the commit message. Instead, maybe just say that an error could occur for unknown reason when a signal is delivered before tasks start (or while they are being started), so we paper over the error by exiting immediately.